src/generic_datatype.cpp

```diff
@@ -0,0 +1,146 @@
+#include "generic_datatype.h"
+
+#include <algorithm>
+#include <exception>
+#include <memory>
+#include <unordered_map>
+
+#include "assert.h"
+
+using namespace std;
+
+namespace smt {
+
+GenericDatatypeDecl::GenericDatatypeDecl(const std::string name) : dt_name(name)
+{
+}
+
+std::string GenericDatatypeDecl::get_name() { return dt_name; }
+
+GenericDatatypeConstructorDecl::GenericDatatypeConstructorDecl(
+    const std::string & name)
+  : cons_name(name),
+    // Should I turn this into a constant?
+    dt_name("NOT IMPLEMENTED YET")
+{
+}
+
+void GenericDatatypeConstructorDecl::add_new_selector(
+    const SelectorComponents & newSelector)
+{
+  // This seems like an opportunity for a new comparison function for
+  // the struct.
```

Comment:
Comments that go into master should only clarify the code, explain variables, etc. TODO comments like that are better outside, except for special cases.